Car docs bot: diff on generated readme file

marcellofuschi opened the pull request:
Closes https://github.com/commaai/openpilot/issues/32690

- Doesn't use pickle, it just compares the markdown files (the one in master and the one in a PR).
- Only the car docs table is compared. The rest of the file is ignored.
- It relies on the (Make, Model) pair from the table to identify a row. If a column value was changed but the pair remained the same, that's shown as a column change in the PR comment (as opposed to an addition+removal).
- Adds the Detail sentence as an expandable button in the docs (like "Hardware Needed").

Here's an example result: https://github.com/marcellofuschi/openpilot/pull/2#issuecomment-2284364319
Detail sentence isn't visible in the example because that column is not in master yet.

marcellofuschi committed:
Change how car docs diff for PR comment is generated

marcellofuschi committed:
Add detail sentence to car docs

marcellofuschi committed:
Fix style

marcellofuschi commented:
Need to fix: the code for parsing the markdown table is too naive and can't handle escaped pipes correctly. Will do it in a bit.

sshane commented:
The sentence is not very easily to read, is there something we can do about that? Tooltip, comment? We don't show it now, so it's not a requirement to show it now.

![image](https://github.com/user-attachments/assets/f8a2b8e5-6bf2-41eb-8790-10fbe4630369)

And does this satisfy this point in the issue?
>Changes to column values should be picked up as a column change, not addition and removal. GitHub already does that for us!

marcellofuschi committed:
Make rendered car docs table more readable

marcellofuschi commented:
>The sentence is not very easy to read, is there something we can do about that?

Pushed a change that should improve that. I automatically add some `&nbsp;` to the column value so that once rendered it prefers overflowing with scroll (up to a certain point).

>And does this satisfy this point in the issue?
>"Changes to column values should be picked up as a column change, not addition and removal."

I think so, as long as the changes aren't made to the columns I use for identifying a row (i.e., `Make` and `Model`).

---

I also added a length limit to the generated comment, so that it won't show more than 10 rows per category, with an indication of the number of hidden rows. I believe the comment would get truncated anyway if too long so it's better to handle it more gracefully.

---

>Need to fix: the code for parsing the markdown table is too naive and doesn't handle potential escaped pipes.

Still gotta fix this.

marcellofuschi committed:
Limit length of the PR comments for car docs diff

marcellofuschi committed:
Fix error in comment generation

marcellofuschi committed:
Prevent trying to create PR comments in the CI when it doesn't have permission

sshane commented:
Getting:

```python
batman@workstation-shane:~/openpilot/selfdrive/debug$ python3 print_docs_diff.py 
  File "/home/batman/openpilot/selfdrive/debug/print_docs_diff.py", line 22
    results = {f'{r['Make']} {r['Model']}': r for r in results}
                     ^^^^
SyntaxError: f-string: unmatched '['
```

sshane commented on selfdrive/debug/print_docs_diff.py:
```diff
@@ -1,120 +1,93 @@
+  match = re.search(r"(\d+)\s+supported\s+cars\s*\n([\s\S]*?\|.*?\|[\s\S]*?\|[-:]+\|[\s\S]*?\n(?:[^\S\r\n]*\S.*\n)*\n?)", docs_content, re.IGNORECASE)
+  if not match:
+    raise RuntimeError("Couldn't find the car docs Markdown table.")
+  num_cars = int(match.group(1))
+  md_table = match.group(2).strip()
```
```suggestion
  num_cars, md_table = match.groups()
```

I like unpacking in one line for the best readability, and you know immediately where the usage of `match` ends.

sshane commented on selfdrive/debug/print_docs_diff.py:
```diff
@@ -1,120 +1,93 @@
-      row_diff = build_column_diff(base_car, new_car)
-      if ARROW_SYMBOL in row_diff:
-        changes["column"].append(row_diff)
+  base_headers, base_cars = get_cars_docs_in_markdown(base_docs_content)
+  new_headers, new_cars = get_cars_docs_in_markdown(new_docs_content)
```
It makes more sense if `new_docs_content` is hardcoded using BASEDIR + CARS.md file locally, and `base_docs_content` should be passed in as a path, using the base commit on the PR (downloaded in the Action)

marcellofuschi replied:
I made the change (https://github.com/commaai/openpilot/pull/33273/commits/ce8c77cdb15874ca3eea2555e4830202a58cfa00) and tested it in my fork. Is this what you had in mind?

sshane commented on selfdrive/debug/print_docs_diff.py:
```diff
@@ -1,120 +1,93 @@
-      if len(changes[category]):
+    for title, category in (("## 🔀 Column Changes", "column"), ("## ❌ Removed", "removals"), ("## ➕ Added", "additions")):
+      ordered_headers = base_headers if category != 'additions' else new_headers
+      max_rows_to_show = 10
```
Unless there is a problem specific to this PR, let's keep this printing behavior the same (no row limit).

marcellofuschi committed:
Attempt to fix cars docs table style

marcellofuschi commented:
> I would say it's okay to hide the sentence for now, it changes the style of the table too much and pushes the other columns near the end off the screen. Another user put it in the markdown file as comments, with some small special handling.

This was partly because I had also enlarged the `Supported Package` column. I reverted it now, so the [table](https://github.com/marcellofuschi/openpilot/blob/car-docs-diff/docs/CARS.md) should look a bit more familiar. Let me know if this could do it.

Regarding `re.sub`s, the easier to remove would be the first one, though I think it's more reliable to also check for word boundary before the appearance of `assets/` when you replace. Let me know if you agree, otherwise I'll change it to `.replace('assets/', 'https://media.githubusercontent.com/media/commaai/openpilot/master/docs/assets/')`

I'll look into the other points you raised (thanks for the feedback).

marcellofuschi committed:
Remove length limitation for car docs PR comment by bot

marcellofuschi committed:
Refactor

marcellofuschi committed:
Refactor

marcellofuschi commented:
>Getting this, why doesn't CI catch it?:

I reproduced the error, I think it has to do with quotes used wrong. Demo example:

<img width="360" alt="Screenshot 2024-08-13 at 4 04 30 PM" src="https://github.com/user-attachments/assets/b031b970-e2bd-44db-8b14-0c191952eaf1">


Fixed it in the code, so it should be okay now. Don't know why CI wasn't catching it.

marcellofuschi committed:
Change how CI calls car docs diff script for PR comment

sshane commented on .github/workflows/selfdrive_tests.yaml:
```diff
@@ -289,9 +289,9 @@ jobs:
           ref: ${{ github.event.pull_request.base.ref }}
```
We want to get rid of the two checkouts in this PR and just use curl/wget

marcellofuschi replied:
Done (https://github.com/commaai/openpilot/pull/33273/commits/ddbe96471e627abb5fe22f0e8df54cd1a9e25260)

marcellofuschi committed:
Avoid multiple checkouts in CI for generating base car docs

marcellofuschi committed:
Handle escaped pipes in car docs table when generating diff in CI

marcellofuschi committed:
Fix unused import

marcellofuschi committed:
Update CARS.md

marcellofuschi commented:
Rebased after the recent changes, and tested again [here](https://github.com/marcellofuschi/openpilot/pull/6)

The Detail sentence is missing from the docs because `docs_definitions.py` was moved to `opendbc`, so the diff doesn't currently involve Detail sentence.

I think it looked fine when we added it to the table as a new column (with formatting done through `&nbsp;` usage). I opened a [PR](https://github.com/commaai/opendbc/pull/1182) to `opendbc` to re-add it.

marcellofuschi committed:
Add test
